### PR TITLE
Complete Ch1

### PR DESCRIPTION
**Pull request by JosephTLucas**
### Solutions for Chapter 1

cc @vrom911 @chshersh

Great material, team!  I know it's not much, but I bought a couple coffees (https://Ko-fi.com/home/coffeeshop?txid=7434d56f-49e4-459b-96ef-a2e85b167254&mode=public&img=ogiboughtsomeone). I can tell you put a lot of time into this and I really appreciate it.

I felt very comfortable with `where` but less so with `let in`. I'll have to do some more research there. Can they always be used interchangeably (with different readability concerns) like `for` and `while` loops in python?  I don't quite have the context yet to understand https://wiki.haskell.org/Let_vs._Where (but hopefully I will after another 3 Chapters).

Thanks again.

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -209,32 +209,31 @@ So, the output in this example means that 'False' has type 'Bool'.
...
 
 >>> :t True
-<INSERT THE RESULT INSTEAD OF THE TEXT>
+Bool
 >>> :t 'a'
-<INSERT THE RESULT INSTEAD OF THE TEXT>
+Char
 >>> :t 42
-<INSERT THE RESULT INSTEAD OF THE TEXT>
+Num
```
Comment (vrom911) [inline src/Chapter1.hs:216]: We haven't covered this in this Chapter as it is a bit more advanced concept, but in reality it will be 
```suggestion
42 :: Num a => a
```
We will explain what it exactly means in Haskell later, but for now, you can think of it as a `Num`eric object. If you look at it this way, it indeed makes sense, as 42 could represent `Int`, `Natural`, `Integer` and even `Double`! So it is a generalised method to say the type of any numeric 🙂 

Btw, we check this example with some test, that is why it requires the exact output of the GHCI on that. Let us know if we could help with running GHCI. If you installed everything that was in the instructions, you should be able to:

```
$ ghci
```

File: src/Chapter1.hs [hunk 2]
```diff
@@ -209,32 +209,31 @@ So, the output in this example means that 'False' has type 'Bool'.
...
 
 A pair of boolean and char:
 >>> :t (True, 'x')
-<INSERT THE RESULT INSTEAD OF THE TEXT>
+(Bool, Char)
 
 Boolean negation:
 >>> :t not
-<INSERT THE RESULT INSTEAD OF THE TEXT>
+Bool
```
Comment (vrom911) [inline src/Chapter1.hs:224]: That is a nice guess!
However, `not` represents the negation of some `Bool`. `not False` is `True` and vice versa.

So by itself, `not` is the _function_, and it takes a `Bool` and returns another `Bool`. In haskell, you would write it like `Bool -> Bool`
```suggestion
not :: Bool -> Bool
```

File: src/Chapter1.hs [hunk 3]
```diff
@@ -209,32 +209,31 @@ So, the output in this example means that 'False' has type 'Bool'.
...
 
 Boolean 'and' operator:
 >>> :t (&&)
-<INSERT THE RESULT INSTEAD OF THE TEXT>
+Bool
```
Comment (vrom911) [inline src/Chapter1.hs:228]: Almost the same in here. `&&` takes two booleans and returns another `Bool`. Can you guess the type now? 🙂

File: src/Chapter1.hs [hunk 4]
```diff
@@ -209,32 +209,31 @@ So, the output in this example means that 'False' has type 'Bool'.
...
 Addition of two numbers:
 >>> :t (+)
-<INSERT THE RESULT INSTEAD OF THE TEXT>
+Num
 
 Maximum of two values:
 >>> :t max
-<INSERT THE RESULT INSTEAD OF THE TEXT>
-
+Ord
```
Comment (vrom911) [inline src/Chapter1.hs:236]: `Ord` will be covered later as well, but to give you a short overview, it is needed to say that you can compare values. And to find the maximum, you need to be able to compare 🙂

File: src/Chapter1.hs [hunk 5]
```diff
@@ -478,7 +478,8 @@ Implement a function that returns the last digit of a given number.
   whether it works for you!
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Int -> Int
+lastDigit n = mod n 10
```
Comment (vrom911) [inline src/Chapter1.hs:482]: Great!
One small note, it won't work correctly on the negative numbers, unfortunately.
Here you would need to use one more standard function – to take the __abs__olute value of `n` 🙂

File: src/Chapter1.hs [hunk 6]
```diff
@@ -541,8 +542,11 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
-
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+    | (x > (min y z)) && (x < (max y z)) = x
+    | (y > (min x z)) && (y < (max x z)) = y
+    | (z > (min x y)) && (z < (max x y)) = z
```
Comment (vrom911) [inline src/Chapter1.hs:549]: This is correct!
However, as we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guard have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.
Because of that, you will need to add another guard – `| otherwise = z`, to tell the compiler, that your pattern matching is exhaustive 🙂

File: src/Chapter1.hs [hunk 7]
```diff
@@ -619,8 +626,13 @@ Implement a function that returns the sum of the last two digits of a number.
...
 -}
-
-sumLast2 n = error "sumLast2: Not implemented!"
+sumLast2 :: Int -> Int
+sumLast2 n = first n + second n
+    where
+        first :: Int -> Int
+        first x = div (mod x 100) 10
+        second :: Int -> Int
+        second x = mod (mod x 100) 10
```
Comment (vrom911) [inline src/Chapter1.hs:635]: That is a wonderful solution! 👏🏼  You correctly noticed that it is the `div` and `mod`, cool 😎 

One hint to make your solution even shorter: the standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that is actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(first, second) = divMod (mod x 100) 10
```

You can see how we could pattern match on the pair 🙂

Comment (vrom911) [reply src/Chapter1.hs:635]: Also, here would also need to be careful on the negative input 👌🏼

File: src/Chapter1.hs [hunk 8]
```diff
@@ -640,8 +652,13 @@ Implement a function that returns the first digit of a given number.
...
 -}
-
-firstDigit n = error "firstDigit: Not implemented!"
+firstDigit :: Int -> Int
+firstDigit n = if n < 10 then n else reduce n
+    where
+        reduce :: Int -> Int
+        reduce x
+            | x < 10 = x
+            | otherwise = reduce (div x 10)
```
Comment (vrom911) [inline src/Chapter1.hs:661]: Even though this is a totally valid way to write it, it could be simplified a bit 🙂 

You can see that you first do

```haskell
if n < 10 
then n 
else reduce n
```

But due to the way of what are guards, it could also be written in this way:

```haskell
  | n < 10 = n 
  | otherwise = reduce n
```

But if you look at the `reduce` function, you can see, that you are doing the same check again, even though you have done it before.

That means, that you can remove the check before running reduce, so you can say just:

```haskell
firstDigit n = reduce n
    where
        reduce :: Int -> Int
...
```
But that means, that the `reduce` function, is completely the same as `firstDigit`. So you can skip introducing the new function in `where` and instead, write the inlined `reduce` body in the `firstDigit` itself (and make it recursive).

Let me know if it makes sense.

Also, negative inputs 🙂

**Review by vrom911 — APPROVED** [review]
That's a fantastic start on Chapter 1 🥳 
And great job on the __Advanced__ task as well!

Thanks a lot for your kind words about the project 😊 We appreciate your support soo much. This really means a world to us ❤️ 

I understand your struggle with `let` `where` scoping. I was struggling with it too, and still will need to spend some extra time on some complicated cases. We decided not to cover the comparison in the course, as it is rather a wide topic. But I am currently working on the blog post exactly about scoping of the variables, which would also explain how `let` and `where` are different in this meaning.

In short, `where` would be available in the whole scope of the function body (including `let`s). But `let` is visible only in the code below in, but not in the wider boundaries than `let` itself. That is why `let` variables are not visible in `where`. Hope it makes some sense. But you should not worry too much about this topic at the moment 🙂

File: src/Chapter1.hs (hunk 8 above)
Comment (JosephTLucas) [reply src/Chapter1.hs:661]: Really great feedback all around, @vrom911! Thanks so much. I'll go back and play with those.  On the first couple type questions:  I actually did run those all in ghci, but didn't yet understand the significance of `x -> x` statements, so was excluding them.  After working through the rest of the Chapter, I understand that concept better.

**Review by vrom911 — APPROVED** [review]
Perfect! I think this is ready to be merged 🚀

**Comment by vrom911** [comment]
⚠️ By the way, per new rules of the Hacktoberfest event, you would need to add the `hacktoberfest` topic to this fork repo:

* https://hacktoberfest.digitalocean.com/hacktoberfest-update

All previous contributions are legit, but from now on it should have the label to participate.
